Make all wasp cli commands to use optparse-applicative

modotte opened the pull request:
### Description

This replaces Wasp CLI custom internal parsers with optparse-applicative for consistency. This also adds automatic *bash*, *zsh* and *fish* shell completions.

fixes #773 

### Command parser coverage progress
| Command                            | Implementation      | Unit Test (Parser)                | Note                                                                                                                                               |
|------------------------------------|---------------------|---------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| `new [name] [-t\|--template]` | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `version`                          | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `waspls [--log LOG_FILE] [--stdio]`                           | <ul><li>- [x] </li> | <ul><li>- [x] </li> | vscode passes --stdio to the language server. waspls always uses --stdio, so this switch is ignored. it's included for vscode compatibility                                                                                                                             |
| `completion [generate (bash\|zsh\|fish)]`                     | <ul><li>- [x] </li> | <ul><li>- [x] </li> | breaking change: use `completion generate shell` format instead of `completion:generate:shell` for consistency with the rest of CLI commands                                     |
| `uninstall`                        | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `start [db]`                       | <ul><li>- [x] </li> | <ul><li>- [x] </li> | we merge both start with and without db options together while maintaining same previous behavior                                                                                                                                              |
| `db (start\|reset\|seed [name]\|studio\|migrate-dev [--name migration-name] [--create-only])`               | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                         |
| `compile`                          | <ul><li>- [x] </li> | <ul><li>- [x] </li> | internally called by `wasp start` - see [discord message](https://discord.com/channels/686873244791210014/1107993093149573220/1107995379452088321) |
| `clean`                            | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `build`                            | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `deploy [args]`                           | <ul><li>- [x] </li> | <ul><li>- [x] </li> | All the arguments will be supplied by external tool (handled by deploy tool of choice, by default it's currently with fly)                                                                      |
| `telemetry`                        | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `deps`                             | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `dockerfile`                       | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `info`                             | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                                    |
| `test (client) [args]`                             | <ul><li>- [x] </li> | <ul><li>- [x] </li> |                                                                                                                                 |

### Misc

1. [x] Double check and fix help prompt messages.
2. [x] Reorganize files (specific command parser code goes into <COMMAND_NAME>/Parser.hs). Do we want to include a simple, single command parser like version and info though?
3. [ ] Restore original command colors if possible. 
4. [x] Rename variables and functions to follow the existing coding style. (not sure about pure vs return)
5. [x] Make unit tests description more meaningful, such as: `db start -t shouldn't pass`.


### Select what type of change this PR introduces:

1. [ ] **Just code/docs improvement** (no functional change). 
6. [ ] **Bug fix** (non-breaking change which fixes an issue).
7. [ ] **New feature** (non-breaking change which adds functionality).
8. [x] **Breaking change** (fix or feature that would cause existing functionality to not work as expected).

### Update Waspc ChangeLog and version if needed
If you did a bug fix, new feature, or breaking change, that affects waspc, make sure you satisfy the following:
1. [x] I updated [ChangeLog.md](https://github.com/wasp-lang/wasp/blob/main/waspc/ChangeLog.md) with description of the change this PR introduces.
3. [ ] I bumped waspc version in [waspc.cabal](https://github.com/wasp-lang/wasp/blob/main/waspc/waspc.cabal) to reflect changes I introduced, with regards to the version of the latest wasp release, if the bump was needed.

infomiho commented on waspc/cli/src/Wasp/Cli/Command/Start.hs:
```diff
@@ -10,19 +10,26 @@ import Control.Monad.Except (throwError)
+start :: StartArg -> Command ()
+start arg = case arg of
+  StartDb -> Wasp.Cli.Command.Start.Db.start
+  StartNormal -> startNormal
```
I would maybe go for more meaningful `startApp`

infomiho commented on waspc/cli/src/Wasp/Cli/Command/WaspLS.hs:
```diff
@@ -3,49 +3,10 @@ module Wasp.Cli.Command.WaspLS
-            <> O.help "Use stdio for communicating with LSP client. This is the only communication method we support for now, so this is the default anyway and this flag has no effect."
-        )
+runWaspLS :: WaspLSArgs -> Command ()
+runWaspLS WaspLSArgs {wslLogFile = lf, wslUseStudio = _} = liftIO $ LS.serve lf
```
Maybe consider full name `logFile` vs. `lf`

infomiho reviewed (commented):
I've left some comments, most of them pretty minor. Good job! I love the fact I can get ZSH completions on my laptop.

I've tested out the CLI locally and everything thing seems to work as expected. The only thing I miss are the yellow command names 😢

modotte commented on waspc/cli/src/Wasp/Cli/Command/Test/Parser.hs:
```diff
@@ -0,0 +1,29 @@
+      O.subparser $
+        mconcat
+          [ mkWrapperCommand "client" CTForwardOptions (TestClient <$> O.many testRestArgs) "Run your app client tests.",
+            mkWrapperCommand "server" CTForwardOptions (TestServer <$> O.many testRestArgs) "Run your app server tests."
```
Sure :+1: ! By hiding, do you mean turning it into an internal/hidden command or we should just drop this command off for now if it's actually not being used?

Martinsos replied:
That's a good question, I think any of that is ok! I do wonder why @shayneczyzewski put it though, maybe the idea was that it is visible, so people can be like "oh look, server tests are coming also"? But we can remove it for now and possibly revert if @shayneczyzewski advises so.

shayneczyzewski replied:
It can be removed if people want to. I mainly added so (a) we had the internal placeholder for `wasp test server` ready and (b) we had some output for the server command that was hopefully friendlier than "command not found" as I figured people may be tempted to try it after running `wasp test client`. But perhaps neither of those warranted it from the start.

Martinsos replied:
Ok, @modotte feel free to do here whatever you like the best and then resolve this comment once you are happy with it.

modotte commented on waspc/cli/src/Wasp/Cli/Command/ShellCompletion.hs:
```diff
@@ -0,0 +1,126 @@
+    unlines
+      [ "Run the following command to generate shell (bash, zsh, or fish) completion script for wasp on your machine:",
+        "",
+        "For Bash completion: ",
```
Tysm! :+1:  I'll incorporate these suggestions in next commits!

modotte commented on waspc/cli/src/Wasp/Cli/Command/Call.hs:
```diff
@@ -1,25 +1,56 @@
-  = New Arguments
-  | Start
-  | StartDb
+  = New !NewArgs
```
Oh right :slightly_smiling_face: ! I just went with what `stan` was suggesting to reduce potential of unexpected space leaks. I'm not 100% sure how this exactly works or whether that's true. I'm more than happy to revert it back from strict if it'll cause any form regression. :smiley:

modotte commented on waspc/cli/src/Wasp/Cli/Command/Call.hs:
```diff
@@ -1,25 +1,56 @@
+  deriving (Show, Eq)
+
+data DbMigrateDevArgs = DbMigrateDevArgs
+  { dbmdaName :: !(Maybe String),
```
I'd prefer to drop the prefix myself, but my primary concern without it is name clashing. However, after a second thought, it shouldn't be too much of an issue with that if we go with prefix-less naming right? :thinking:

modotte replied:
I went with renaming them to something more readable instead as suggested here https://github.com/wasp-lang/wasp/pull/1202#discussion_r1202503046

modotte commented on waspc/cli/src/Wasp/Cli/Parser/Util.hs:
```diff
@@ -0,0 +1,22 @@
+import Options.Applicative (Parser)
+import qualified Options.Applicative as O
+
+data CommandType = CTNormal | CTNoIntersperse | CTForwardOptions
```
Thank you! :+1:  Hmm, that's true with the prefix and modifiers. This actually prompts me re-think this a bit because optparse-applicative have [more](https://hackage.haskell.org/package/optparse-applicative-0.18.0.0/docs/Options-Applicative.html#t:InfoMod) info modifiers provided than by my new `CommandType`. 

This current code forces the code user to update the type and the `withModifier` internal function if they want to add different modifier in the future. So, I decided to pass a list of modifiers instead.

What I'm thinking I wanna go for right now is:

```haskell

mkWrapperCommand :: String -> [InfoMod a] -> Parser a -> String -> O.Mod O.CommandFields a
mkWrapperCommand name infoModifiers callCommand description =
  O.command
    name
    ( O.info
        (O.helper <*> callCommand)
        (O.progDesc description <> mconcat infoModifiers)
    )
```

This way, they can add more than one modifiers if they choose to do so and doesn't need to update multiple things at once.

> Another question, are we sure we need both and not just forward options?

You're correct! I think we only need `forwardOptions`. My reasoning for `deploy` command to use `noIntersperse` was to reject something like `wasp deploy --flag` early on because I don't see anything like this being used in the [docs](https://wasp-lang.dev/docs/deploying). 

I also thought if `deploy` provided with arguments, it need to start with at least, one tool name (`wasp deploy deployToolName --blah something etc -v`).

Martinsos replied:
Aha, well
> This is the same behaviour one sees after the "--" pseudo-argument.
this sounds good.

Ok, I will let you investigate this a bit more, but it would make sense to me if we can have this special behaviour only for specific commands, for which we know that they forward their arguments. If we can't fine, that is survivable, but if we can that sounds cleaner.

modotte replied:
@martinsos sorry. :sweat_smile:  I believe I've read your question a bit wrong. By this:

> Turn them on only for wasp deploy though and disable them for the rest of the CLI

I think it already works this way for deploy. Currently, only `wasp test client` and `wasp deploy` use `forwardOptions` to treat unknown arguments as positionals (we could make this even looser with `allPositional`, which I think makes more sense if we really want `node` to handle ***all*** arguments for these two commands).

The rest of the CLI commands (such as new, etc) won't be affected by this. 

btw, I think we should use `allPositional` for deploy.

Examples of `forwardOptions` vs `allPositional` for `deploy`:

## forwardOptions
This will still exit and print our own optparse-applicative help message if we pass help flag.
```sh
$ wasp-cli deploy --help

Usage: wasp-cli deploy [DEPLOY_ARGUMENTS]

  Deploys your Wasp app to cloud hosting providers.

Available options:
  -h,--help                Show this help text
  DEPLOY_ARGUMENTS         Currently only supports fly.io. See
                           https://wasp-lang.dev/docs/deploying.

```

## allPositional
This will pass ***all*** arguments after `wasp deploy` to node. This is also the exact same behavior before this PR.
```sh
$ wasp-cli deploy --help

Usage: wasp deploy [options] [command]

CLI for deploying Wasp apps to various clouds

Options:
  -h, --help      display help for command

Commands:
  fly             Create and deploy Wasp apps on Fly.io
  help [command]  display help for command

```

Martinsos replied:
Ok sounds great that it is already limited to just those two commands, and very nice analysis here @modotte -> all sounds reasonable, allPositional seems like obvious winner here then! You understand this part better than me at this point so whatever you find better, I am ok with.

Martinsos commented on waspc/cli/src/Wasp/Cli/Command/CreateNewProject.hs:
```diff
@@ -26,10 +23,8 @@ import Wasp.Cli.Common (WaspProjectDir)
-createNewProject args = do
-  newProjectArgs <- parseNewProjectArgs args & either throwProjectCreationError return
+createNewProject :: NewArgs -> Command ()
+createNewProject newProjectArgs = do
```
I was looking at `NewArgs` here, wondering where is the right place to define them, and ended up thinking generally about how we define Commands in our CLI.

So a bit of general thinking! Let's say we want to add a new command to Wasp CLI. We would need to define following stuff:
1. Command parser (for optparse) -> ok, this might be trivial, so it is just `pure Smth`, but it can be a proper parser also.
2. Command Call.
3. Command args type. Sometimes there are none, sometimes it is just list of Strings or something simple, but often it is a custom type (record).
4. Command logic that actually executes the command (in Command monad).
5. Name and usage/description for a command.

Some relations: Args (3) are used by Call (2) and command logic (4). Parser (1) generates Call (2), so it also transitively uses Args (3). Parser (1) and name and usage (5) are combined together to form the final entry for optparse, to represent the command.

Ok, that was a bit of analysis!

The question is, where do we put all these parts, in the codebase? Right now Parser (1) is sometimes next to logic (4) or is directly or is directly in optparse definition of the whole CLI. Name and usage (5) is sometimes together with Parser (1), sometimes it is directly in the optparse definition in the CLI. Args (3) are right now defined next to Command Call (2).

One thing we can improve for sure here is ensure things are consistent. Main inconsistency right now is that name and usage are sometimes in specific command Parser, but sometimes they are directly in the top level Parser -> I would make sure they are always in the top level Parser.

Next, it would be interesting to explore if we could collocate things more.
What if every command would have all its parts grouped?
So we would have all parts, from 1 to 5, in the same module, be it in multiple sub-modules or in one sub-module.
So for example, for command New, we would have its logic in one file, parser in file next to it, args maybe in the same file as logic, name and usage possibly in the same file as parser like you were doing now. I am not sure only about Command Call -> that one maybe we could leave outside, where it is now. Yeah, I think so.

I am not 100% sure about this, it would require a bit of playing and seeing what it looks, but I think something in this direction might be interesting. If you feel it is too much for this PR, we can possibly leave this as the very last thing to play with, or even for the future PR. One thing I would address is just making sure that name and usage are always in the same place.

Martinsos replied:
@modotte I am curious what you think after reading this -> if you have a good idea how to improve some of the stuff here and want to try out, that could be interesting.
If you don't and any direction you pick looks like more trouble than it is worth, then I am completely fine with leaving it as it is for now.

modotte replied:
I don't really have a clear idea of how to do this, unfortunately. Maybe we should leave this one in the future/separate issue/for a different PR? 

Although, I'd really like the idea of bringing the relevant things closer together for cohesiveness. That certainly would make it easier if someone else wants to add/remove/update the commands in the future.

Martinsos replied:
Yup, the collocation would be nice! I am ok with leaving this as it is for now then, unless you want to try something out. So I will leave it to you, feel free to resolve this comment if you think it would be better to leave this for the future (which makes sense), or keep it open if you have some ideas / approaches and we can work on that.

Martinsos reviewed (changes requested):
Great job @modotte !
I think this is really good all together -> you did a nice job introducing opt parse, you modified the code around it to make sense, there is nothing special that is off -> looks pretty good!

I did leave quite a few comments, but mostly they are not abour arhitecture but about naming.
There is one about arhitecture and how we organize each command, that is one a bit more interesting and also not so well defined, so we can discuss that one further.

Looking forward to your responses!

Martinsos reviewed (changes requested):
Great job @modotte !
I think this is really good all together -> you did a nice job introducing opt parse, you modified the code around it to make sense, there is nothing special that is off -> looks pretty good!

I did leave quite a few comments, but mostly they are not abour arhitecture but about naming.
There is one about arhitecture and how we organize each command, that is one a bit more interesting and also not so well defined, so we can discuss that one further.

Looking forward to your responses!

Martinsos commented:
One convention we try to use @modotte is that comments are resolved only by the person that started them. This way we can be sure that both sides were happy with the resolution.

What we also do, is that author of the PR leaves an emoji on the comment to confirm they saw it / solved it. We use :+1: to annotate that comment was soon and taken into account, and we use :rocket: to indicate that comment was solved (code change was commited and pushed).

If these rules are too much now we can adapt, but in case you are ok with following them, they might make our reviewing process somewhat easier :)!

Martinsos commented on waspc/cli/exe/Main.hs:
```diff
@@ -95,49 +47,28 @@ main = withUtf8 . (`E.catch` handleInternalErrors) $ do
     handleInternalErrors :: E.ErrorCall -> IO ()
     handleInternalErrors e = putStrLn $ "\nInternal Wasp error (bug in compiler):\n" ++ indent 2 (show e)
 
-printUsage :: IO ()
```
I also found this conversation, maybe it is useful: https://github.com/pcapriotti/optparse-applicative/issues/88 .

Might also be worth asking directly on GH repo of optparse (opening an issue), and/or asking on r/haskell.

Martinsos replied:
And this section in their README is what we care about: https://github.com/pcapriotti/optparse-applicative#customising-the-help-screen . But it doesn't say much about stuff like coloring hm.

modotte replied:
I've opened an issue in optparse-applicative repo https://github.com/pcapriotti/optparse-applicative/issues/478. One of the maintainers has replied the issue. I don't think it's possible at the moment. Hmm. :thinking: :crying_cat_face:. Maybe I could look further into https://github.com/wasp-lang/wasp/pull/1202#discussion_r1204735326 and see what I could find for a workaround.

Martinsos replied:
Ha that sucks -> I commented on the issue asking for some ideas on how to go about it, if we could maybe create a PR.

Martinsos replied:
Btw there was this comment, referring to (COMMAND | COMMAND) thing that is happening right now: https://github.com/pcapriotti/optparse-applicative/issues/478#issuecomment-1571824106 -> we should also fix that!

Martinsos replied:
Idea: what if we put ANSI codes for color, bold and similar, directly into the names of subgroups and commands while providing them to optparse? Ah yes that won't work because it uses those names not only to show them but also to parse them, damn. Ok maybe it would work for subgroups actually hm, but it will certainly not work for commands.

modotte commented on waspc/cli/src/Wasp/Cli/Parser.hs:
```diff
@@ -0,0 +1,64 @@
+    options =
+      O.info (parserSuite <**> O.helper) $ O.footerDoc (Just footer)
+
+parserSuite :: Parser Call
```
Oh! I take inspiration from `testSuite` commonly used to group bunch of smaller tests together. We could perhaps name it something like `parseGroup` or similar, maybe?

modotte replied:
:thinking: Hmm. alright, I'll try. 

>This function handles three set of subparsers. Each of these subparsers are in charge of parsing all top level commands. Based on first CLI arg (new, compile, db, etc), it will call either subparser to handle the rest of CLI args.

With this, I think it could be `topLevelCommandsParser`, perhaps? 

Also, after reading your thoughts about the suffix, I decided to rename the parsers with `Parser` suffix. Personally, I think this makes more sense because it's more consistent with the parser return type `Parser a`.

Pushed these changes to 52bbc514343b65888ffb87a941d827bc9913a813.

Martinsos replied:
Both sounds good, nice!

modotte commented on waspc/cli/src/Wasp/Cli/Command/ShellCompletion.hs:
```diff
@@ -0,0 +1,126 @@
+        "then reset your terminal session."
+      ]
+
+-- TODO: Use newly exposed`*shCompletionScript` functions from optparse-applicative >0.17
```
There was a recent patch release from optparse-applicative (0.17.1.0) which now exposes these three functions. I think we could drop them and use from optparse instead. However, there's a blockade from current  pinned `ormolu` to use optparse 0.17.0 I believe?

If that's the case, is it alright if we bump ormolu version the newest release (0.7.0.0)?

modotte replied:
After a little bit playing around with the dependencies, I found out that we can stick with existing ormolu. What's causing the issue on my end is that I didn't update `index-state` in `dev-tool.project`.

Martinsos replied:
ah uff, that is actually probably a mistake on our side, we shoudl have update that! Sorry about that.

Martinsos replied:
So you managed to use newer optparse and we now don't need these functions? Feel free to resolve this comment if so.

modotte commented on waspc/cli/src/Wasp/Cli/Command/Db/Parser.hs:
```diff
@@ -0,0 +1,50 @@
+        mconcat
+          [ mkCommand "start" (pure DbStart) "Alias for `wasp start db`.",
+            mkCommand "reset" (pure DbReset) "Drops all data and tables from development database and re-applies all migrations.",
+            -- FIXME: Fix stdout formatting.
```
Please see https://github.com/wasp-lang/wasp/pull/1202#discussion_r1204407617. Tbf, I should place the fixme closer to the problem.

modotte commented on waspc/cli/src/Wasp/Cli/Command/Start.hs:
```diff
@@ -10,19 +10,26 @@ import Control.Monad.Except (throwError)
-start :: Command ()
-start = do
+startNormal :: Command ()
+startNormal = do
```
I went with `startApp` (and StartApp for the type) instead as suggested by @infomiho and you. Do let me know if it's probably much better to rename this to something else.

modotte commented on waspc/cli/src/Wasp/Cli/Command/ShellCompletion.hs:
```diff
@@ -0,0 +1,126 @@
+    Shell (Bash, Fish, Zsh),
+  )
+
+completion :: CompletionArgs -> Command ()
+completion args = case args of
+  ShowInstruction -> printShellCompletionInstruction
+  Generate shell -> generateSh $ case shell of
+    Bash -> bashCompletionScript
+    Zsh -> zshCompletionScript
+    Fish -> fishCompletionScript
```
This is certainly much clearer! ty!

modotte commented on waspc/cli/src/Wasp/Cli/Parser.hs:
```diff
@@ -0,0 +1,64 @@
+parserSuite = internalCommands <|> generalCommands <|> inProjectCommands
+
+internalCommands :: Parser Call
+internalCommands = O.subparser $ mkCommand "compile" (pure Compile) "" <> O.internal
```
Hi, can you please explain a bit more on what do you mean by skipping them? :smiley: 

As in hiding `compile` command whenever `--help/-h` is invoked or no arguments are being passed? If that's the case, `O.internal` should do just that or do you meant we should move it into a subparser file reserved for internal commands?

Martinsos replied:
Ah yes I meant not showing them in help -> I didn't realize O.internal takes care of that! Awesome then :).

modotte commented on waspc/cli/src/Wasp/Cli/Parser.hs:
```diff
@@ -0,0 +1,64 @@
+            Term.applyStyles [Term.Cyan] "Newsletter:" <> " https://wasp-lang.dev/#signup"
+          ]
+    options =
+      O.info (parserSuite <**> O.helper) $ O.footerDoc (Just footer)
```
Got it! Ty for the newspaper analogy too. I'll try to apply this to my other code as well.

modotte commented on waspc/cli/src/Wasp/Cli/Parser/Util.hs:
```diff
@@ -0,0 +1,22 @@
+
+data CommandType = CTNormal | CTNoIntersperse | CTForwardOptions
+
+mkWrapperCommand :: String -> CommandType -> Parser a -> String -> O.Mod O.CommandFields a
```
Oh, I've deprecated this a bit earlier to go for (I think) more flexible and simple approach as discussed here -https://github.com/wasp-lang/wasp/pull/1202#discussion_r1204503271.

but yeah, I should probably rename the functions as suggested too.

modotte commented on waspc/cli/src/Wasp/Cli/Command/Db/Migrate.hs:
```diff
@@ -33,25 +33,14 @@ migrateDev optionalMigrateArgs = do
-    go unknown _ = Left $ "Unknown migrate arg(s): " ++ unwords unknown
+      let migrateArgs DbMigrateDevArgs {dbmdaName = name, dbmdaCreateOnly = isCreateOnly} =
+            MigrateArgs {_isCreateOnlyMigration = isCreateOnly, _migrationName = name}
+      ExceptT $ DbOps.migrateDevAndCopyToSource dbMigrationsDir projectRootDir $ migrateArgs optionalMigrateArgs
```
:+1: Ty! Yes, we don't need runExceptT and ExceptT anymore here. optparse-applicative and its based parsers should handle all of these errors for us now.

modotte commented on waspc/cli/src/Wasp/Cli/Command/Db/Parser.hs:
```diff
@@ -0,0 +1,50 @@
+        mconcat
+          [ mkCommand "start" (pure DbStart) "Alias for `wasp start db`.",
+            mkCommand "reset" (pure DbReset) "Drops all data and tables from development database and re-applies all migrations.",
+            -- FIXME: Fix stdout formatting.
```
This has been fixed in https://github.com/wasp-lang/wasp/pull/1202/commits/6d7ec5f422f3fb19114c4f5d4079182ce4062f9d.

Martinsos reviewed (commented):
@modotte looking good, I resolved a lot of comments, and commented more on the rest -> give it another look please!

Martinsos reviewed (commented):
@modotte looking good, I resolved a lot of comments, and commented more on the rest -> give it another look please!

modotte commented on waspc/cli/src/Wasp/Cli/Command/ShellCompletion.hs:
```diff
@@ -0,0 +1,159 @@
+completion PrintInstruction = printShellCompletionInstruction
+completion (PrintScript shell) = printShellCompletionScript shell
+
+type ShellCompletionScript = (String -> String -> String)
```
Oops! My mistake. Revealing the names does help expose a bug. :sweat_smile:  I've found out that `progn` is the running program name and `prog` is the `progn` absolute filepath. 

If I understand this correctly, current code implementation completion script won't work in situation where the running program is not in `PATH` because the actual completer need to invoke the waspc exe for optparse-applicative own internal completion flags such as `--bash-completion-word`, etc. This won't work if it can't find exe location! :fearful: 

For example, from optparse-applicative completion README: ```$(foo --bash-completion-script `which foo`)``` is pretty much equals to `bashCompletionScript foo fooAbsolutePath`.


So perhaps what I should do to fix:

1. First and foremost check if we have existing utility in waspc/cli to get the absolute path of current running program, if not I'll implement it on my own with `findExecutable` and StrongPath.
2. Replace `prog` value with the actual absolute filepath.


For names, I might go with these for now:
```haskell
type AbsProgFile = FilePath
type ProgName = ProgName
type ShellCompletionScriptBuilder = (AbsProgFile -> ProgName -> String)
```

modotte replied:
Pushed the fix into https://github.com/wasp-lang/wasp/pull/1202/commits/34aa3beaebbb043e2deb148e6468fcf5d7c06245

Martinsos replied:
Ok, certainly an improvement, and nice think catching a bug! Ah yes, names reveal much :D.

I would maybe go with `AbsProgPath` or `AbsProfFilePath` instead, but otherwise sounds good. Also providing a comment above ShellCompletionScriptBuilder explaining the relationship between AbsProgFile and ProgName would be helpful -> because those are referring to the same program but it might not be obvious.

modotte commented on waspc/cli/src/Wasp/Cli/Parser.hs:
```diff
@@ -0,0 +1,67 @@
+import Wasp.Cli.Parser.Util (mkCommand, mkNormalCommand)
+import qualified Wasp.Util.Terminal as Term
+
+parseCliArgs :: IO Call
```
I've renamed the Call type to CommandCall as suggested, but I left the `Wasp.Cli.Command.Call` module name as it is for now. I intended to update it to reflect the type as well, but I think `Wasp.Cli.Command.CommandCall` for module name doesn't offer any extra clarity. Should I leave it alone? What do you think? :thinking:

modotte commented on waspc/cli/src/Wasp/Cli/Parser.hs:
```diff
@@ -0,0 +1,64 @@
+        start,
+        db,
+        mkCommand "clean" (pure Clean) "Deletes all generated code and other cached artifacts.",
+        mkCommand "build" (pure Build) "Generates full web app code, ready for deployment. Use when deploying or ejecting.",
```
Pushed this suggestion into: a8282b6de92e2a4e99747caae0ac150e1357e02c. Sorry for the wait on this.

Martinsos reviewed (commented):
@modotte I reviewed your new changes!

There are some comments you haven't addressed yet, without counting the big one with refactoring that we are leaving for the last.
If you are not sure which ones are still open, go through `Conversations` and you should find them.

We are coming close to an end I think, so I would love us to wrap this up and merge it! If any of open comments seems like too much work / troublesome, do pls comments so and we can figure out what to do with it.

modotte commented on waspc/cli/src/Wasp/Cli/Command/Call.hs:
```diff
@@ -1,25 +1,56 @@
+  | Test !TestArgs -- "client" | "server", then test cmd passthrough args
+  deriving (Show, Eq)
+
+data Shell = Bash | Zsh | Fish deriving (Show, Eq)
```
Hmm, if we do this, I think we should move `CompletionArgs` closer to Completion stuff too. Maybe, it's better if we move all top-level arguments type closer to their parser or command runner (into their own type module) for consistency? :thinking:

modotte replied:
Hmm, if we do this, I think we should move `CompletionArgs` closer to Completion stuff too. Maybe, it's better if we move all top-level arguments type closer to their parser or command runner (into their own type module) for consistency. :thinking:

modotte replied:
Alright. I've tried with moving the `Shell` data type into `Wasp.Cli.Command.ShellCompletion.Parser`, however this would introduce cyclic dependency with `Wasp.Cli.Command.Call` because `CompletionArgs` in the module is depending on `Shell` as well. This also applies if I move `Shell` into `Wasp.Cli.Command.ShellCompletion`. Both of these parser and logic runner modules are depending on `CommandCall`.

So I think what I perhaps could do:

1. Leave `Shell` in `Wasp.Cli.Command.Call` as it is.
-or-
2. Introduce a new module `Wasp.Cli.Command.ShellCompletion.CompletionArgs` and bring both `Shell` and `CompletionArgs` into it.
    - Make `Wasp.Cli.Command.Call` and Completion modules import them normally.
    - I've tried 2. and it works all right (with ParserTest too), but I find it odd to see only CompletionArgs in their own specific module. So, I wonder if we should move the rest of command args type to their own module as well, following the same approach.

Martinsos replied:
Ha, this is getting a bit complex!
And this really ties in with the other big comment I left, about deciding where these data structures go.

We gave it a try, but since it is not clear, maybe it is best to just leave it as it is, until with time it becomes clearer how we can handle it. What do you think?

modotte replied:
Sure, we can jump directly to the big comment and solve this one over there.

Martinsos replied:
We can maybe even also give up on the big comment, that is what I was trying to suggest :D.

modotte replied:
Ohh! Okay then sure. :)

modotte commented on waspc/cli/src/Wasp/Cli/Command/Db/Cli.hs:
```diff
@@ -0,0 +1,20 @@
+import Wasp.Cli.Command.Db.Studio (studio)
+import Wasp.Cli.Command.Start.Db (start)
+
+dbCli :: DbArgs -> IO ()
```
Sorry! I think I'll go with renaming `runDbCommand` to `runAsDbCommand` and `dbCli` to `runDbCommand` as suggested. Certainly better than `dbCli`, I believe.

Martinsos replied:
Oh nice, that is great naming! Go for it!

modotte commented on waspc/cli/src/Wasp/Cli/Command/CreateNewProject/Parser.hs:
`Mod` is the option modifier. In our case, the command will return command fields along with its modifier so we can, if we want to concat/add new behavior to the command such as turning it into an internal command, grouping it into specific commandGroup (like our `inProjectCommands`), chain it with another regular command, or with any other valid modifier that could be applied to [it](https://hackage.haskell.org/package/optparse-applicative-0.17.1.0/docs/Options-Applicative-Builder.html#v:hidden). 

One name that I came up with above to make the type a bit more clear is this:
`CommandWithOptionModifier`

Martinsos replied:
Hm I am not sure because I have not studied optparse enough, but it seems to me like their names are not the best :D.
What we can do is potentially use some kind of type aliases like this, although we hide a bit of info -> if somebody knows optparse they will be expecting the `Mod` stuff.

Maybe best to leave it as it is then, I am not sure if we can really improve it much -> we can give it richer names, but still it will be confusing unless you understand what a modifier is, and once you do you will probably also understand the `Mod` signature.

Let me know what you think but I am ok with leaving it as it is, because I don't think there is a clear and simple improvement here.

modotte replied:
If that's the case then sure, we'll leave this one as it is. :slightly_smiling_face:

modotte commented on waspc/cli/src/Wasp/Cli/Parser/Util.hs:
```diff
@@ -0,0 +1,16 @@
+import qualified Options.Applicative as O
+
+mkCommand :: String -> [InfoMod a] -> Parser a -> Mod CommandFields a
+mkCommand name infoModifiers commandCall =
```
> If you are assuming that a in the signature will be a Command.Call or something like that, then we might want to capture that in the type system.

Hmm :thinking:, but we also use this for some command args that are not part of `CommmandCall` such as `Parser WaspLSArgs`, and on primitives including bool `Parser Bool`, and string `Parser String`. Not sure what should I do here :point_right::point_left: 

> Also, what is this Mod CommandFiels a? How would you explain to me what that is? I am looking at it and I don't know how to reason about it, what to call it in my head. I guess it is not just optparse parser, so what is it?

I left a comment In here - https://github.com/wasp-lang/wasp/pull/1202#discussion_r1214022722

Martinsos replied:
Aha no worries, if it indeed is an `a` then leave it as that! That is ok, what I was suggesting is that if it is always the same, if it is always commandCall (as arg name indicated), then we could express that in the type.
But then please also rename `commandCall` to `commandParser`, and that should help.

Martinsos replied:
Just a reminder for this and other comments: put a :rocket: when you are done, so I don't have to go checking the code and can just resolve the comment.

Martinsos replied:
Hehe thanks for all the rockets, they help a lot!

Martinsos commented on waspc/cli/src/Wasp/Cli/Command/Deploy/Parser.hs:
```diff
@@ -0,0 +1,27 @@
+import Wasp.Cli.Command.Call (Call (Deploy))
+import Wasp.Cli.Parser.Util (CommandType (CTNoIntersperse), mkWrapperCommand)
+
+-- The rest of arguments will be supplied externally by chosen deploy tool.
```
Ok great! Once you remove it, feel free to resolve this comment.

modotte commented on waspc/cli/test/Wasp/Cli/ParserTest.hs:
```diff
@@ -0,0 +1,77 @@
+                ("info", Info)
+              ]
+
+    describe "new should correctly pass with" $ do
```
Went with https://github.com/wasp-lang/wasp/pull/1202#discussion_r1210276234 instead

modotte commented:
@Martinsos Hi again. Thx for all the reviews. I've addressed most of the comments including the `(COMMAND | COMMAND)` help output mentioned yesterday. I've also restored `EXAMPLES` in front of the main help page (I forgot to put it back, and it just makes sense to have it back as a guide for new users anyways).

The last big one (architecture) and the color of the commands (which is probably better in the long-term if solved directly in optparse-applicative) are the major ones that are still left to be solved.

Martinsos commented:
Awesome @modotte and thanks for this nice summary, it helps a lot with reviewing!

- Big refactoring -> I commented on it -> up to you, you can resolve it if you wish! I will let you make the final call.
- Colors -> well, this one is tricky. I am a bit hesitant to merge this PR until we actually have that solved, since we would regress in not having nice visuals. I agree it would be best to get support for it from optparse, that is a long term solution. If that doesn't work though, do you have any ideas for a hacky solution that could work? Can we intercept the string it returns for help and modify it / decorate it, even if with somewhat fragile logic?

modotte commented:
Alright! I'll take a look at what I can do in the meantime to get the command colors back again. :slightly_smiling_face:

Martinsos commented:
> Alright! I'll take a look at what I can do in the meantime to get the command colors back again. slightly_smiling_face

I just took a quick look myself -> doesn't feel like much can be done? I continued discussion on the optparse repo, I think the only way through this is that we make a PR there to implement this.

Martinsos commented:
Since coloring is main thing stopping us from merging this PR, I took a bit of time to play with trying to extend optparse-applicative to support cusotmization of help, here is some very early draft: https://github.com/pcapriotti/optparse-applicative/pull/482 .
